### PR TITLE
Fix gradient accumulation in SFT training loop

### PR DESCRIPTION
**Pull request by svlandeg**
Alternative to https://github.com/karpathy/nanochat/pull/59, implementing & testing @kibitzing's suggestion from https://github.com/karpathy/nanochat/pull/59#issuecomment-3408851450.

Currently, the loss is normalized by dividing by `grad_accum_steps`. However, the number of meaningful tokens per mini batch may vary, and so it's mathematically more correct to use `loss_reduction='sum'` instead of `'mean'` and then divide by the total number of meaningful tokens (`num_tokens`) in the full batch.

To verify, I've run some quick experiments, logging the sum of the parameter weights as 
```
param_sums = sum([x.sum().item() for x in model.parameters()])
```
This should be roughly the same for the same initial model, SFT data & seed, irrespective of the size of the mini batch (`device_batch_size`).

## Master

Single mini batch:

```
Target examples per step: 60
Device batch size: 60
=> Setting grad accum steps: 1

INIT Model param sums 473866.49

STEP 0 : num tokens 22421
 - MICRO STEP 0 : len 60  / num tokens 22421
   Model param sums 474003.83

STEP 1 : num tokens 21786
 - MICRO STEP 0 : len 60  / num tokens 21786
   Model param sums 474097.06
```

Mini batch of size 10:

```
Target examples per step: 60
Device batch size: 10
=> Setting grad accum steps: 6

INIT Model param sums 473866.49

STEP 0 : num tokens 22421
 - MICRO STEP 0 : len 10  / num tokens 4121
 - MICRO STEP 1 : len 10  / num tokens 3564
 - MICRO STEP 2 : len 10  / num tokens 4055
 - MICRO STEP 3 : len 10  / num tokens 3378
 - MICRO STEP 4 : len 10  / num tokens 3124
 - MICRO STEP 5 : len 10  / num tokens 4179
   Model param sums 474005.63

STEP 1 : num tokens 21786
 - MICRO STEP 0 : len 10  / num tokens 3320
 - MICRO STEP 1 : len 10  / num tokens 3073
 - MICRO STEP 2 : len 10  / num tokens 3889
 - MICRO STEP 3 : len 10  / num tokens 3529
 - MICRO STEP 4 : len 10  / num tokens 3788
 - MICRO STEP 5 : len 10  / num tokens 4187
   Model param sums 474100.38
```
Note the discrepancy in step 0 (474003.83 vs 474005.63) and step 1 (474097.06 vs 474100.38)

## This PR

Single mini batch:

```
Target examples per step: 60
Device batch size: 60
=> Setting grad accum steps: 1

INIT Model param sums 473866.49

STEP 0 : num tokens 22421
 - MICRO STEP 0 : len 60  / num tokens 22421
   Model param sums 474003.83

STEP 1 : num tokens 21786
 - MICRO STEP 0 : len 60  / num tokens 21786
   Model param sums 474097.06
```

Mini batch of size 10:

```
Target examples per step: 60
Device batch size: 10
=> Setting grad accum steps: 6

INIT Model param sums 473866.49

STEP 0 : num tokens 22421
 - MICRO STEP 0 : len 10  / num tokens 4121
 - MICRO STEP 1 : len 10  / num tokens 3564
 - MICRO STEP 2 : len 10  / num tokens 4055
 - MICRO STEP 3 : len 10  / num tokens 3378
 - MICRO STEP 4 : len 10  / num tokens 3124
 - MICRO STEP 5 : len 10  / num tokens 4179
   Model param sums 474003.88

STEP 1 : num tokens 21786
 - MICRO STEP 0 : len 10  / num tokens 3320
 - MICRO STEP 1 : len 10  / num tokens 3073
 - MICRO STEP 2 : len 10  / num tokens 3889
 - MICRO STEP 3 : len 10  / num tokens 3529
 - MICRO STEP 4 : len 10  / num tokens 3788
 - MICRO STEP 5 : len 10  / num tokens 4187
   Model param sums 474097.13
```

Note that the results are much closer: step 0 (474003.83 vs 474003.88) and step 1 (474097.06 vs 474097.13) with the small deviation probably due to floating point errors.

## PR 59

Single mini batch:

```
Target examples per step: 60
Device batch size: 60
=> Setting grad accum steps: 1

INIT Model param sums 473866.49

STEP 0 : num tokens 22421
 - MICRO STEP 0 : len 60  / num tokens 22421
   Model param sums 474003.83

STEP 1 : num tokens 21786
 - MICRO STEP 0 : len 60  / num tokens 21786
   Model param sums 474096.99
```

Mini batch of size 10:

```
Target examples per step: 60
Device batch size: 10
=> Setting grad accum steps: 6

INIT Model param sums 473866.49

STEP 0 : num tokens 22421
 - MICRO STEP 0 : len 10  / num tokens 4121
 - MICRO STEP 1 : len 10  / num tokens 3564
 - MICRO STEP 2 : len 10  / num tokens 4055
 - MICRO STEP 3 : len 10  / num tokens 3378
 - MICRO STEP 4 : len 10  / num tokens 3124
 - MICRO STEP 5 : len 10  / num tokens 4179
   Model param sums 474003.88

STEP 1 : num tokens 21786
 - MICRO STEP 0 : len 10  / num tokens 3320
 - MICRO STEP 1 : len 10  / num tokens 3073
 - MICRO STEP 2 : len 10  / num tokens 3889
 - MICRO STEP 3 : len 10  / num tokens 3529
 - MICRO STEP 4 : len 10  / num tokens 3788
 - MICRO STEP 5 : len 10  / num tokens 4187
   Model param sums 474097.13
```

The results are very similar to this PR: step 0 (474003.83 vs 474003.88) and step 1 (474096.99 vs 474097.13) with the only difference being the value for step 1 single-batch (474097.06 on `master` and this PR, 474096.99 for PR 59).

### CODE REVIEW

**Review by svlandeg — COMMENTED** [review]
Functional comparison with https://github.com/karpathy/nanochat/pull/59 ,which adjusts the gradients afterwards by running `param.grad.div_(num_tokens_item)`:

- IMO the implementation on this PR is more "clean" & understandable
- The main disadvantage of this PR is that the training data is fully loaded in memory at L214. This could be remedied by looping through it and determining `num_tokens` op front, then releasing the memory again and looping through the data once more from scratch for the `microstep` training loop at L221.

File: scripts/chat_sft.py [hunk 1]
```diff
@@ -211,17 +211,19 @@ def get_lr_multiplier(it):
...
+    num_tokens += sum((targets >= 0).sum() for _, targets in steps)
+
+    if ddp:
+        dist.all_reduce(num_tokens, op=dist.ReduceOp.SUM) # sum over ranks
+
+    for micro_step, (train_inputs, train_targets) in enumerate(steps):
         with autocast_ctx:
-            loss = model(train_inputs, train_targets)
+            loss = model(train_inputs, train_targets, loss_reduction='sum')
+        loss = loss / num_tokens # normalize loss here
```
Comment (kibitzing) [inline scripts/chat_sft.py:229]: I'd suggest including the safety check for `num_tokens == 0`. It prevents `NaN` loss in those rare edge cases where a batch has no valid tokens.

Comment (kibitzing) [reply scripts/chat_sft.py:229]:
```python
if num_tokens.item() == 0:
    model.zero_grad(set_to_none=True)
    continue
```

Comment (svlandeg) [reply scripts/chat_sft.py:229]: I left this out originally so that the PR would only focus on the actual bug fix. Have you ever seen this happen? I would think something is properly wrong upstream if a batch has no tokens?

Comment (kibitzing) [reply scripts/chat_sft.py:229]: I haven’t personally seen it happen either. It’s more of a defensive thing on my side. Whenever there’s a division involved, I get a bit nervous without a safety check.

I agree that it might indicate something is wrong upstream, but I think it could also happen in some edge cases due to the data itself or truncation. That said, you’re right that if we’re already taking a mean, having zero valid tokens would result in NaN anyway, so leaving it as-is is probably fine too.

Comment (svlandeg) [reply scripts/chat_sft.py:229]: Yea. I definitely see where you're coming from, I just wanted to do the most minimal PR for hopefully maximum clarity and faster mergeability 😉  Thanks for reviewing! 🙏
